hedera-node/docs/proposals/block-stream-bucket-uploader.md

```diff
@@ -0,0 +1,398 @@
+├── /uploaded/     # Successfully uploaded blocks
+└── /hashmismatch/ # Blocks with hash validation failures
+```
+
+### Extended BlockStreamWriterMode
+
+```java
+public enum BlockStreamWriterMode {
+    GRPC,   // Stream blocks via gRPC
+    FILE,   // Write blocks to local files
+    BUCKETS // Write blocks to cloud storage buckets
+}
+```
+
+### Configuration
+
+The configuration is split into two parts:
+1. Network properties that can be updated at runtime
+2. Secure credentials stored locally
+
+#### Network Properties
+`@NetworkProperty` Are network-wide properties that can be dynamically updated via file 0.0.121.
+The `writerMode` node property exists to support the ability for each node to be able to choose which modes it will use to write blocks.
+
+```java
+@ConfigData("blockStream")
+public record BlockStreamConfig(
+    // ... existing code ...
+    // New Network Properties
+    @ConfigProperty(defaultValue = "3") @NetworkProperty int uploadRetryAttempts,
+    @ConfigProperty(defaultValue = "168") @NetworkProperty int localRetentionHours,
+    @ConfigProperty(defaultValue = "data/config/bucket-credentials.json") @NetworkProperty String credentialsPath,
+    // Bucket configurations with default AWS and GCP public buckets
+    @ConfigProperty(defaultValue = """
+        [
+            {
+                "name": "default-aws-bucket",
+                "provider": "aws",
+                "endpoint": "https://s3.amazonaws.com",
+                "region": "us-east-1",
+                "bucketName": "hedera-mainnet-blocks"
+            },
+            {
+                "name": "default-gcp-bucket",
+                "provider": "gcp",
+                "endpoint": "https://storage.googleapis.com",
+                "region": "",
+                "bucketName": "hedera-mainnet-blocks"
+            }
+        ]
+        """) @NetworkProperty List<BucketNetworkConfig> buckets,
+    // Updated Node Property for writerModes
+    @ConfigProperty(defaultValue = "[FILE, BUCKETS]") @NodeProperty List<BlockStreamWriterMode> writerModes,
+    // ... existing code ...
+) {}
+
+public record BucketNetworkConfig(
+    @ConfigProperty String name,
+    @ConfigProperty String provider,  // "aws" or "gcp"
+    @ConfigProperty String endpoint,
+    @ConfigProperty String region,    // required for AWS only
+    @ConfigProperty String bucketName
+) {
+    public BucketNetworkConfig {
+        Objects.requireNonNull(name, "Bucket name cannot be null");
+        Objects.requireNonNull(provider, "Provider cannot be null");
+        Objects.requireNonNull(endpoint, "Endpoint cannot be null");
+        Objects.requireNonNull(bucketName, "Bucket name cannot be null");
+        
+        if ("aws".equals(provider)) {
+            Objects.requireNonNull(region, "Region is required for AWS buckets");
+        }
+    }
+}
+```
+
+#### Credentials Configuration
+The credentials are stored in a separate JSON file on disk for security:
+
```

Comment:
As per @nathanklick we can use configuration for credentials management but it will need to be symmetric key encrypted on disk (possibly secured with the gossip signing key)